sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala

```diff
@@ -652,6 +654,152 @@ case class StringRPad(str: Expression, len: Expression, pad: Expression)
+  // from UTF8String to String to URL for every row.
+  @transient private lazy val cachedUrl = children(0) match {
+    case Literal(url: UTF8String, _) => if (url ne null) getUrl(url) else null
+    case _ => null
+  }
+
+  // If the key is a constant, cache the Pattern object so that we don't need to convert key
+  // from UTF8String to String to StringBuilder to String to Pattern for every row.
+  @transient private lazy val cachedPattern = children(2) match {
+    case Literal(key: UTF8String, _) => if (key ne null) getPattern(key) else null
+    case _ => null
+  }
+
+  // If the partToExtract is a constant, cache the Extract part function so that we don't need
+  // to check the partToExtract for every row.
+  @transient private lazy val cachedExtractPartFunc = children(1) match {
+    case Literal(part: UTF8String, _) => getExtractPartFunc(part)
+    case _ => null
+  }
+
+  import ParseUrl._
+
+  override def checkInputDataTypes(): TypeCheckResult = {
+    if (children.size > 3 || children.size < 2) {
+      TypeCheckResult.TypeCheckFailure(s"$prettyName function requires two or three arguments")
+    } else {
+      super[ImplicitCastInputTypes].checkInputDataTypes()
+    }
+  }
+
+  private def getPattern(key: UTF8String): Pattern = {
+    Pattern.compile(REGEXPREFIX + key.toString + REGEXSUBFIX)
+  }
+
+  private def getUrl(url: UTF8String): URL = {
+    try {
+      new URL(url.toString)
+    } catch {
+      case e: MalformedURLException => null
+    }
+  }
+
+  private def getExtractPartFunc(partToExtract: UTF8String): (URL) => String = {
+    partToExtract match {
+      case HOST => (url: URL) => url.getHost
+      case PATH => (url: URL) => url.getPath
+      case QUERY => (url: URL) => url.getQuery
+      case REF => (url: URL) => url.getRef
+      case PROTOCOL => (url: URL) => url.getProtocol
+      case FILE => (url: URL) => url.getFile
+      case AUTHORITY => (url: URL) => url.getAuthority
+      case USERINFO => (url: URL) => url.getUserInfo
+      case _ => (url: URL) => null
+    }
+  }
+
+  private def extractValueFromQuery(query: UTF8String, pattern: Pattern): UTF8String = {
+    val m = pattern.matcher(query.toString)
+    if (m.find()) {
+      UTF8String.fromString(m.group(2))
+    } else {
+      null
+    }
+  }
+
+  private def extractFromUrl(url: URL, partToExtract: UTF8String): UTF8String = {
+    if (cachedExtractPartFunc ne null) {
+      UTF8String.fromString(cachedExtractPartFunc.apply(url))
+    } else {
+      UTF8String.fromString(getExtractPartFunc(partToExtract).apply(url))
+    }
+  }
+
+  private def parseUrlWithoutKey(url: UTF8String, partToExtract: UTF8String): UTF8String = {
+    if (cachedUrl ne null) {
+      extractFromUrl(cachedUrl, partToExtract)
+    } else {
+      val currentUrl = getUrl(url)
+      if (currentUrl ne null) {
+        extractFromUrl(currentUrl, partToExtract)
+      } else {
+        null
+      }
+    }
+  }
+
+  override def eval(input: InternalRow): Any = {
```

Comment:
this is somewhat convoluted with 4 levels of nesting, I think you can rewrite it this way to make it easier to follow
```
val evaluated = children.map(_.eval(input).asInstanceOf[UTF8String])
if (evaluated.contains(null)) {
  return null
}

if (evaluated.size == 2) {
  return parseUrlWithoutKey(evaluated(0), evaluated(1))
} else {
  // 3-arg, i.e. QUERY with key
  assert(evaluated.size == 3)
  if (evaluated(1) != QUERY) {
    return null
  }

  val query = parseUrlWithoutKey(evaluated(0), evaluated(1))
  if (query eq null) {
    return null
  }

  if (cachedPattern ne null) {
    return extractValueFromQuery(query, cachedPattern)
  } else {
    return extractValueFromQuery(query, getPattern(evaluated(2)))
  }
}
```